src/utils/Hardware.ts

```diff
@@ -0,0 +1,246 @@
+}
+
+/**
+ * Implements the `Hardware` API by accessing it through a web service.
+ */
+export class WebServiceHardware implements Hardware {
+  /**
+   * Reads Accessible Controller status
+   */
+  public async readAccesssibleControllerStatus(): Promise<
+    AcessibleControllerAPI
+  > {
+    return await fetchJSON<AcessibleControllerAPI>(
+      '/hardware/accessible_controller/read_status'
+    )
+  }
+  /**
+   * Sets Accessible Controller connected
+   */
+  public async setAccesssibleControllerConnected(
+    connected: boolean
+  ): Promise<void> {
+    await fetch('/hardware/accessible_controller/set_connected', {
+      method: 'post',
+      body: JSON.stringify(connected),
+    })
+  }
+
+  /**
+   * Reads Battery status
+   */
+  public async readBatteryStatus(): Promise<BatteryAPI> {
+    return await fetchJSON<BatteryAPI>('/hardware/battery/read_status')
+  }
+
+  /**
+   * Sets Battery charging
+   */
+  public async setBatteryCharging(charging: boolean): Promise<void> {
+    await fetch('/hardware/battery/set_charging', {
+      method: 'post',
+      body: JSON.stringify(charging),
+    })
+  }
+
+  /**
+   * Sets Battery level. Number between 0–1.
+   */
+  public async setBatteryLevel(level: number): Promise<void> {
+    await fetch('/hardware/battery/set_level', {
+      method: 'post',
+      body: JSON.stringify(level),
+    })
+  }
+
+  /**
+   * Reads Card Reader status
+   */
+  public async readCardReaderStatus(): Promise<CardReaderAPI> {
+    return await fetchJSON<CardReaderAPI>('/hardware/card_reader/read_status')
+  }
+
+  /**
+   * Sets Card Reader connected
+   */
+  public async setCardReaderConnected(connected: boolean): Promise<void> {
+    await fetch('/hardware/card_reader/set_connected', {
+      method: 'post',
+      body: JSON.stringify(connected),
+    })
+  }
+
+  /**
+   * Reads Printer status
+   */
+  public async readPrinterStatus(): Promise<PrinterAPI> {
+    return await fetchJSON<PrinterAPI>('/hardware/printer/read_status')
+  }
+
+  /**
+   * Sets Card Reader connected
+   */
+  public async setPrinterConnected(connected: boolean): Promise<void> {
+    await fetch('/hardware/printer/set_connected', {
+      method: 'post',
+      body: JSON.stringify(connected),
+    })
+  }
+}
+
+/**
+ * Implements the `Hardware` API with an in-memory implementation.
+ */
+export class MemoryHardware implements Hardware {
+  private accessibleControllerStatus: AcessibleControllerAPI = {
+    connected: true,
+  }
+  private batteryStatus: BatteryAPI = {
+    charging: true,
+    level: 0.8,
+  }
+  private cardReaderStatus: CardReaderAPI = {
+    connected: true,
+  }
+  private printerStatus: PrinterAPI = {
+    connected: true,
+  }
+
+  /**
+   * Reads Accessible Controller status
+   */
+  public async readAccesssibleControllerStatus(): Promise<
+    AcessibleControllerAPI
+  > {
+    console.log("Why is 'this' undefined?", this)
+    return this.accessibleControllerStatus
```

Comment:
@eventualbuddha - The last commit on this branch is a WIP with debugging code. I'm stumped as to why `this` is undefined. `MemoryHardware` is modeled after the `MemoryCard`, but something must be different. Do you see where the error in this code is?

Reply:
Commented at the source of the problem. It’s an unbound method problem. I’d fix it by doing `const { hardware } = this.props` and calling i.e. `await hardware.read…`.